Mathlib/GameTheory/Auction/Basic.lean

```diff
@@ -0,0 +1,192 @@
+Authors: Ma Jiajun, Wang Haocheng
+-/
+import Mathlib.Data.Fintype.Basic
+import Mathlib.Data.Finset.Lattice
+import Mathlib.Data.Real.Basic
+
+/-!
+# Auction Theory
+
+This file formalizes core concepts and results in auction theory. It includes the definitions of
+first-price and second-price auctions, as well as several fundamental results and helping lemmas.
+
+## Main Definitions
+
+- `maxBid`: Function that computes the highest bid given a bidding function.
+- `winner`: Identifies the winner of the auction as the bidder with the highest bid.
+- `maxBidExcluding`: Function that computes the highest bid excluding a given participant.
+- `Dominant`: Defines a dominant strategy in the context of a auction.
+- `secondPrice`: Computes the second highest bid in the auction.
+- `Secondprice.utility`: Computes the utility of each bidder based on the outcome of the auction.
+- `Firstprice.utility`: Computes the utility for a first price auction.
+
+## Main Results
+
+- `utility_nneg`: Utility is non-negative if the bid equals the valuation in second price auction.
+- `valuation_is_dominant`: Bidding one's valuation is a dominant strategy in second price auction.
+- `first_price_has_no_dominant_strategy`: There is no dominant strategy in first price auction.
+
+## Helping Lemmas
+
+- `eq_winner_of_bid_gt`: If `i`'s bid is higher than all other bids, then `i` wins.
+- `exists_maxBid`: There exists a participant whose bid matches the highest bid
+- `bid_winner_eq_maxBid`: The winner's bid is the highest.
+- `Secondprice.utility_winner`: If `i` wins, utility is the valuation minus the second highest bid.
+- `Secondprice.utility_loser`: If `i` does not win, their utility is 0.
+- `bid_le_bid_winner`: The winner's bid is greater than or equal to all other bids.
+- `maxBidExcluding_eq_maxBid_if_loser`: If `i` does not win, the highest bid excluding `i` equals
+   the highest bid.
+- `Firstprice.utility_winner`: If `i` wins in a first price auction utility is valuation minus bid.
+- `Firstprice.utility_loser`: If `i` does not win in a first price auction, utility is 0.
+
+## Implementation Notes
+
+The structure and functions assume the existence of multiple bidders to allow for meaningful
+auction dynamics. Definitions like `winner` and `maxBid` make use of Lean's `Finset` and `Classical`
+logic to handle potential non-constructive cases effectively.
+
+## References
+
+* [T. Roughgarden, *Twenty lectures on Algorithmic Game Theory*][roughgarden2016]
+
+## Tags
+
+auction, game theory, economics, bidding, valuation
+-/
+
+/- `v` is the valuation for participants `I` -/
+variable {I : Type*} [Fintype I] [Nontrivial I] (v : I → ℝ)
+
+namespace Auction
+
+/- `b` is the bidding function that maps participant to a real number bid. -/
+variable (b : I → ℝ)
+
+/-!### Helper Functions and Definitions -/
+/-- Computes the highest bid given a bidding function `b`. -/
+@[simp]
+def maxBid : ℝ := Finset.sup' Finset.univ Finset.univ_nonempty b
```

Comment:
... this is a bit idiosyncratic and if you are to include this in mathlib then you shouldn't make this definition